### PR TITLE
Improve `#to_str` and `#to_json` for `Version::NULL`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Mentioned this here: https://github.com/Homebrew/brew/pull/15402#discussion_r1190824114

### CODE REVIEW

File: Library/Homebrew/version.rb [hunk 1]
```diff
@@ -703,7 +703,25 @@ def to_i
   def to_s
     version.to_s
   end
-  alias to_str to_s
+
+  sig { returns(String) }
+  def to_str
+    raise NoMethodError, "undefined method `to_str' for #{self.class}:NULL" if null?
```
Comment (reitermarkus) [inline Library/Homebrew/version.rb:709]: This is kinda hacky, would probably make more sense to deprecate `Version#to_str` instead.

Comment (Bo98) [reply Library/Homebrew/version.rb:709]: Sounds tricky to deprecate given implicit conversions - can't really search to see what does that and there's likely cases in formulae. And all over place really for stuff like: `Pathname("...")/version`

But doing what you're doing here for `NULL` does seem worthwhile to stop any subtle bugs.

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/version.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/version.rb:709]: For the `Pathname("...")/version` case we could actually implement `Version#to_path` instead.

Comment (Bo98) [reply Library/Homebrew/version.rb:709]: Isn't exactly much of an improvement over `to_str` though.

Comment (MikeMcQuaid) [reply Library/Homebrew/version.rb:709]: > Sounds tricky to deprecate given implicit conversions

Agreed 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @reitermarkus!

**Review by carlocab — APPROVED** [review]

File: Library/Homebrew/test/version_spec.rb [hunk 2]
```diff
@@ -256,25 +313,25 @@
 
   describe "::create" do
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply Library/Homebrew/test/version_spec.rb:314]: Yeah, this was just a find-replace. Should say `::new`, since `Version::create` will be deprecated.

**Comment by nandahkrishna** [comment]
@reitermarkus For some reason this commit seems to be causing a weird CI failure on Linux (at Homebrew/homebrew-core#135761), or so `git bisect` tells me. We have:
```
$ brew install -sdv sbcl
...
/usr/bin/env tar --extract --no-same-owner --file /home/linuxbrew/.cache/Homebrew/downloads/d7b3eb2dadc3855f0103ab9789d3ab39e733dc7f17bac5b9c62f2c8bb1ebd2f2--sbcl-2.3.4-source.tar.bz2 --directory /tmp/d20230704-32234-1gq982l
/usr/bin/env cp -pR /tmp/d20230704-32234-1gq982l/sbcl-2.3.4/. /tmp/sbcl-20230704-32234-1j56edt/sbcl-2.3.4
==> Temporary files retained at:
/tmp/sbcl-20230704-32234-1j56edt
Error: An exception occurred within a child process:
  TypeError: no implicit conversion of MacOSVersion into String
```
And I'm unable to debug the exact cause of this. Do you think you could look into this or point me to a fix?

**Comment by Bo98** [comment]
`ENV["SBCL_MACOSX_VERSION_MIN"] = MacOS.version if OS.mac?` should fix it (or `.to_s` but I prefer the `if` given that's the direction we're generally going).

Technically it's a compatibility break yeah. Though it's almost impossible to deprecate the old behaviour and we seem to have survived a few weeks without many noticing.

**Comment by nandahkrishna** [comment]
Ah alright. I thought it should be fixed in `Homebrew/brew` but I'll just edit the Formula for now. Thanks!

**Pull request merged**